File: clients/OpenSync.Client.Js/src/OpenSyncClient.ts
import { Transport, TransportType, WebSocketTransport } from './Connection/Transport';
import { SseTransport } from './Connection/SseTransport';
import { EventEmitter } from './Events/EventEmitter';
import { TokenManager } from './Auth/TokenManager';
import { SyncDocument } from './SyncDocument';
import { SyncList } from './SyncList';
import { SyncMap } from './SyncMap';
import { SyncStream } from './SyncStream';
import { SyncChannel } from './SyncChannel';

export interface SyncEvent {
  event: string;
  object_type: string;
  object_id: string;
  data?: any;
  timestamp: string;
}

export class OpenSyncClient {
  private transport: Transport;
  private emitter: EventEmitter;
  private tokenManager: TokenManager;
  private baseUrl: string;
  private connected: boolean = false;

  constructor(baseUrl: string, tokenManager: TokenManager, transportType: TransportType = TransportType.WebSocket) {
    this.baseUrl = baseUrl;
    this.tokenManager = tokenManager;
    this.emitter = new EventEmitter();

    switch (transportType) {
      case TransportType.SSE:
        this.transport = new SseTransport(baseUrl, tokenManager);
        break;
      default:
        this.transport = new WebSocketTransport(baseUrl, tokenManager);
    }

    this.transport.onMessage((data: string) => {
      try {
        const event: SyncEvent = JSON.parse(data);
        this.emitter.emit(event.event, event);
        this.emitter.emit('*', event);
      } catch (e) {
        console.error('Failed to parse event:', e);
      }
    });
  }

  async connect(): Promise<void> {
    await this.transport.connect();
    this.connected = true;
  }

  async disconnect(): Promise<void> {
    await this.transport.disconnect();
    this.connected = false;
  }

  async subscribe(objectType: string, objectId: string): Promise<void> {
    await this.send({ action: 'subscribe', object_type: objectType, object_id: objectId });
  }

  async unsubscribe(objectType: string, objectId: string): Promise<void> {
    await this.send({ action: 'unsubscribe', object_type: objectType, object_id: objectId });
  }

  async send(message: any): Promise<void> {
    await this.transport.send(JSON.stringify(message));
  }

  on(event: string, handler: (event: SyncEvent) => void): void {
    this.emitter.on(event, handler);
  }

  off(event: string, handler: (event: SyncEvent) => void): void {
    this.emitter.off(event, handler);
  }

  document(name: string): SyncDocument {
    return new SyncDocument(this, name);
  }

  list(name: string): SyncList {
    return new SyncList(this, name);
  }

  map(name: string): SyncMap {
    return new SyncMap(this, name);
  }

  stream(name: string): SyncStream {
    return new SyncStream(this, name);
  }

  channel(name: string): SyncChannel {
    return new SyncChannel(this, name);
  }

  async httpGet<T>(path: string): Promise<T> {
    const token = await this.tokenManager.getToken();
    const response = await fetch(`${this.baseUrl}/api/v1/sync${path}`, {
      headers: { Authorization: `Bearer ${token}` }
    });
    return response.json();
  }

  async httpPost<T>(path: string, body: any): Promise<T> {
    const token = await this.tokenManager.getToken();
    const response = await fetch(`${this.baseUrl}/api/v1/sync${path}`, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${token}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(body)
    });
    return response.json();
  }
}

File: nupkgs/OpenSync.Sdk/OpenSyncClient.cs
using OpenSync.Sdk.Auth;
using OpenSync.Sdk.Events;
using OpenSync.Sdk.Serialization;
using OpenSync.Sdk.Transports;

namespace OpenSync.Sdk;

public class OpenSyncClient : IDisposable
{
    private readonly string _baseUrl;
    private readonly ITokenProvider _tokenProvider;
    private readonly ITransport _transport;
    private readonly JsonEventSerializer _serializer;
    private readonly HttpClient _httpClient;
    private string? _connectionId;

    public event EventHandler<SyncEventArgs>? OnEvent;
    public event EventHandler<EventArgs>? OnConnected;
    public event EventHandler<EventArgs>? OnDisconnected;

    public OpenSyncClient(string baseUrl, ITokenProvider tokenProvider, TransportType transportType = TransportType.WebSocket)
    {
        _baseUrl = baseUrl.TrimEnd('/');
        _tokenProvider = tokenProvider;
        _serializer = new JsonEventSerializer();
        _httpClient = new HttpClient();
        _transport = TransportFactory.Create(transportType, _baseUrl, _serializer, tokenProvider);
        _transport.OnMessage += HandleMessage;
        _transport.OnConnected += (s, e) => OnConnected?.Invoke(this, e);
        _transport.OnDisconnected += (s, e) => OnDisconnected?.Invoke(this, e);
    }

    private void HandleMessage(object? sender, string json)
    {
        var evt = _serializer.DeserializeEvent(json);
        if (evt != null)
            OnEvent?.Invoke(this, evt);
    }

    public async Task ConnectAsync(CancellationToken cancellationToken = default)
    {
        var token = await _tokenProvider.GetTokenAsync();
        await _transport.ConnectAsync(token, cancellationToken);
        _connectionId = Guid.NewGuid().ToString("N");
    }

    public async Task DisconnectAsync()
    {
        await _transport.DisconnectAsync();
        _connectionId = null;
    }

    public async Task SubscribeAsync(string objectType, string objectId, CancellationToken cancellationToken = default)
    {
        var msg = _serializer.Serialize(new { action = "subscribe", object_type = objectType, object_id = objectId });
        await _transport.SendAsync(msg, cancellationToken);
    }

    public async Task UnsubscribeAsync(string objectType, string objectId, CancellationToken cancellationToken = default)
    {
        var msg = _serializer.Serialize(new { action = "unsubscribe", object_type = objectType, object_id = objectId });
        await _transport.SendAsync(msg, cancellationToken);
    }

    public Task<SyncDocument> GetDocumentAsync(string uniqueName)
        => Task.FromResult(new SyncDocument(this, uniqueName));

    public Task<SyncList> GetListAsync(string uniqueName)
        => Task.FromResult(new SyncList(this, uniqueName));

    public Task<SyncMap> GetMapAsync(string uniqueName)
        => Task.FromResult(new SyncMap(this, uniqueName));

    public Task<SyncStream> GetStreamAsync(string uniqueName)
        => Task.FromResult(new SyncStream(this, uniqueName));

    public Task<SyncChannel> GetChannelAsync(string uniqueName)
        => Task.FromResult(new SyncChannel(this, uniqueName));

    public async Task<T> HttpGetAsync<T>(string path)
    {
        var response = await _httpClient.GetAsync($"{_baseUrl}/api/v1/sync{path}");
        response.EnsureSuccessStatusCode();
        var json = await response.Content.ReadAsStringAsync();
        return _serializer.Deserialize<T>(json);
    }

    public async Task<T> HttpPostAsync<T>(string path, object body)
    {
        var content = new StringContent(_serializer.Serialize(body), System.Text.Encoding.UTF8, "application/json");
        var response = await _httpClient.PostAsync($"{_baseUrl}/api/v1/sync{path}", content);
        response.EnsureSuccessStatusCode();
        var json = await response.Content.ReadAsStringAsync();
        return _serializer.Deserialize<T>(json);
    }

    internal async Task SendAsync(string message, CancellationToken cancellationToken = default)
        => await _transport.SendAsync(message, cancellationToken);

    public void Dispose()
    {
        _transport?.DisconnectAsync().GetAwaiter().GetResult();
        _httpClient?.Dispose();
    }
}
